archtec_playerdata/init.lua

```diff
@@ -0,0 +1,396 @@
+local save_interval = 60
+local floor = math.floor
+local time = os.time
+local debug_mode = false
+
+minetest.register_on_mods_loaded(function()
+    if not minetest.mkdir(datadir) then
+		error("[archtec_playerdata] Failed to create datadir directory '" .. datadir .. "'!")
+	end
+end)
+
+-- struct: add new keys with default/fallback values! (Set always 0 as fallback!)
+local struct = {
+    nodes_dug = 0,
+    nodes_placed = 0,
+    items_crafted = 0,
+    died = 0,
+    playtime = 0,
+    chatmessages = 0,
+    joined = 0,
+    join_count = 0
+}
+
+-- helper funtions
+local function log_action(message)
+    if message ~= nil and message ~= "" then
+        minetest.log("action", "[archtec_playerdata] " .. message)
+    end
+end
+
+local function log_warning(message)
+    if message ~= nil and message ~= "" then
+        minetest.log("warning", "[archtec_playerdata] " .. message)
+    end
+end
+
+local function log_debug(message)
+    if debug_mode then
+        if message ~= nil and message ~= "" then
+            minetest.log("warning", "[archtec_playerdata] " .. message)
+        end
+    end
+end
+
+local function valid_player(name)
+    if name ~= nil and name ~= "" and type(name) == "string" then
+        log_debug("valid_player: '" .. name .. "' is valid!")
+        return true
+    else
+        log_warning("valid_player: '" .. name .. "' is not valid!")
+        return false
+    end
+end
+
+local function stats_file_exsist(name)
+    if not valid_player(name) then return end
+    local file = io.open(datadir .. "/" .. name .. ".txt", "r")
+    if file ~= nil then
+        file:close()
+        log_debug("stats_file_exsist: file of '" .. name .. "' exsits")
+        return true
+    else
+        log_debug("stats_file_exsist: file of '" .. name .. "' does not exsit")
+        return false
+    end
+end
+
+local function in_struct(key)
+    return struct[key] ~= nil
+end
+
+local function is_valid(value)
+    local valtype = type(value)
+    if valtype == "number" or valtype == "string" or valtype == "boolean" then
+        return true
+    else
+        return false
+    end
+end
+
+local function divmod(a, b)
+    if type(a) ~= "number" or type(b) ~= "number" then
+        return
+    end
+    return floor(a / b), a % b
+end
+
+local function format_duration(seconds)
+	local display_hours, seconds_left = divmod(seconds, 3600)
+	local display_minutes, display_seconds = divmod(seconds_left, 60)
+	return ("%02d:%02d:%02d"):format(display_hours, display_minutes, display_seconds)
+end
+
+local function in_cache(name)
+    if not valid_player(name) then return false end
+    if cache[name] ~= nil then
+        return true
+    else
+        return false
+    end
+end
+
+local function get_session_playtime(name)
+	if playtime_current[name] then
+		return time() - playtime_current[name]
+	else
+		return 0
+	end
+end
+
+-- save data
+local function stats_save(name)
+    if not valid_player(name) then return end
+    local data = cache[name]
```

Comment:
It's probably not a huge deal, but because the playtime only gets updated from `stats_save_all` the actual saved time will always be a multiple of `save_interval` (in this case minute only). 

This is also a bit of a display issue as playtime doesn't even get updated when its requested to display to user by the `/stats` command, so user will also never see it update real time.

Without a bit of a big restructuring of the code, you could fix it by updating the total playtime both in this call, right before getting cache, and also updating playtime when someone runs the /stats command.

Reply:
I don't think this is a big issue